src/beatree/leaf/store.rs

```diff
@@ -0,0 +1,225 @@
+use crate::beatree::leaf::free_list::FreeList;
+use crate::beatree::leaf::{LeafPage, PageNumber};
+use crate::io::{CompleteIo, IoCommand, IoKind};
+use crate::store::{Page, PAGE_SIZE};
+use crossbeam_channel::{Receiver, Sender, TrySendError};
+
+use std::fs::File;
+use std::os::fd::AsRawFd;
+use std::os::unix::fs::MetadataExt;
+
+pub struct LeafStore {
+    store_file: File,
+    io_handle_index: usize,
+    io_sender: Sender<IoCommand>,
+    io_receiver: Receiver<CompleteIo>,
+    // monotonic page number, used when the free list is empty
+    nonce_page_number: PageNumber,
+    // this is the max supported page number, after which write fails,
+    // `ftruncate` of the file, to make it bigger, must be handled by upper layers
+    max_page_number: PageNumber,
+    free_list: FreeList,
+}
+
+impl LeafStore {
+    /// creates a LeafStore over a possibly already existing File
+    pub fn create(
+        fd: File,
+        free_list_head: Option<PageNumber>,
+        nonce_page_number: PageNumber,
+        io_handle_index: usize,
+        io_sender: Sender<IoCommand>,
+        io_receiver: Receiver<CompleteIo>,
+    ) -> LeafStore {
+        let file_size = fd
+            .metadata()
+            .expect("Error extracting metadata from LeafStore file")
+            .size() as usize;
+
+        let max_page_number = PageNumber((file_size / PAGE_SIZE) as u32);
```

Comment:
I am confused, why are `nonce_page_number` and `max_page_number` different? 

I had assumed assumed that in case the free list is empty, the allocation requests are going to be served from the end of the file (`max_page_number`). Now, I am left wondering, what is `nonce_page_number` anyway?

Reply:
The main difference between `nonce_page_number` and `max_page_number` lies in their association with the `update_size` function. I will explain both concepts here.

In the proposed implementation we discussed, there is an upper layer responsible for writing pages, performing ftruncate if necessary, and then writing the remaining pages.

The key point is that ftruncate does not allocate space for a specific number of pages, rather, it allocates memory in chunks. Thus, `nonce_page_number` represents the current index of the dump, while `max_page_number` represents the maximum `nonce_page_number` that can be stored in the file without requiring a new ftruncate operation. This distinction is necessary because we aim to move the ftruncate operation elsewhere, so the `LeafStore` cannot handle all tasks independently while maintaining only a single dump index.

The `update_size` function comes into play after a commit on `LeafStoreTx`, where the output includes pages that can be written immediately (`to_allocate`) and those that need ftruncate before being written (`exceeded`). After a new `ftruncate` the `LeafStore` must be advertised of the changed file size.

This could be done automatically maybe looking for the new size each `start_tx`